<center><img src = 'https://github.com/insaid2018/Term-1/blob/master/Images/INSAID_Full%20Logo.png?raw=true'  width="400" height="200"></center>

# Data Visualization of INSAIDTelecom Customers Behavioral Data


## Table of Contents

1. [Project Description](#section1)<br>
     
2. [Consulting Goals](#section2)<br>
     
3. [Data Description](#section3)<br>     

4. [Problem Statement](#section4)<br> 

5. [Loading Package and Data](#section5)<br> 
  




5. [Data Pre Processing](#section5)<br> 

6. [EDA of Pre Processed Data](#section6)<br> 

7. [Profiling](#section7)<br> 

8. [Post Profiling EDA Analysis](#section8)<br> 

9. [Conclusion/Summarization](#section9)<br>

10. [Actionable Insight](#section10)<br>

<a id=section1></a>
## 1. Project Description 

- It's always wonderful to see services customized to your needs.
  Businesses try to understand your behavior and adjust their offerings
  so as to ensure you feel attached to their services.


  **InsaidTelecom**, one of the leading telecom players, understands that customizing offering
  is very important for its business to stay competitive.
  Currently, **InsaidTelecom** is seeking to leverage behavioral data from more than 60% of the 50 million mobile devices         active daily in India to help its clients better understand and interact with their audiences.

  In this **consulting assignment** 
  Insaidians are expected to build a dashboard to understand
  user's demographic characteristics based on their mobile usage, geolocation, and mobile device properties.
  Doing so will help millions of developers and brand advertisers around the world pursue
  data-driven marketing efforts which are relevant to their users and catered to their preferences.

<a id=section2></a>
## 2. Consulting Goals 

- To help the customer the consultants are expected to have depth of clarity in the underlying data.
  How much effort has been put into cleansing and purifying the data will decide how closely have you looked at the data.
  How detailed is the observation stated in the submission report and finally how well a group presents their consulting   
  journey.

  Please remember this is an analytics consulting hence, your efforts in terms of
  finding user behavior is going to directly impact the company's offerings.
  Do help the company understand what is the
  right way forward and suggest actionable insights from marketing and product terms.

<a id=section3></a>
## 3. Data Description

- In this assignment, you are going to study the demographics of a user (gender and age) based on their app download and usage   behaviors.
  The Data is collected from mobile apps that use InsaidTelecom services. Full recognition and consent from individual user of   those apps have been obtained,
  and appropriate anonymization have been performed to protect privacy. Due to confidentiality, we won't provide details on how   the gender and age data was obtained.
  Please treat them as accurate ground truth for prediction. The data schema can be represented in the following table:


- Gender_age_train - Devices and their respective user gender, age and age_group

  Phone_brand_device_model - device ids, brand, and models phone_brand: note that few brands are in Chinese

  
  | Brand Name    | Brand English Mapping|
  | ------------- |:-------------       :| 
  | '华为'        |'Huawei'              |
  | '小米'        |'Xiaomi'              |  
  | '三星'        |'Samsung'             | 
  |'vivo'         |'vivo'                |   
  | 'OPPO'        |'OPPO'                |
  | '魅族'        |'Meizu'               |
  | '酷派'        |'Coolpad'             |
  | '乐视'        |'LeEco'               |
  | '联想 '       |'Lenovo'              |
  | 'HTC'	      |'HTC'	             |

  
  **events_data** - when a user uses mobile on **INSAID Telecom network**, the event gets logged in this data.
  Each event has an **event id, location (lat/long)**, and the event corresponds to frequency of mobile usage.
  timestamp: when the user is using the mobile.




<a id=section4></a>
### 4. Problem Statement

**The goal is to _________to be discussed within the team__.**

<a id=section5></a>
### 5. Loading Packages and Data

**Installing required packages and importing library**

In [1]:
#installing mysql connector
!pip install mysql.connector

#Installation of Geopy
!pip install geopy

#Installing Google Translator to translate phone brands name from non-English to English
!pip install googletrans

     |████████████████████████████████| 11.9MB 2.7MB/s 
  Created wheel for mysql.connector: filename=mysql_connector-2.2.9-cp36-cp36m-linux_x86_64.whl size=247949 sha256=605b01f46895a732dc1b6a75160a6d39495cc9e6ec13de5a416beaea9f94ff7e
  Stored in directory: /root/.cache/pip/wheels/8c/83/a1/f8b6d4bb1bd6208bbde1608bbfa7557504bed9eaf2ecf8c175
Successfully built mysql.connector
     |████████████████████████████████| 61kB 2.9MB/s 
     |████████████████████████████████| 51kB 6.2MB/s 
     |████████████████████████████████| 921kB 8.5MB/s 
     |████████████████████████████████| 61kB 8.2MB/s 
     |████████████████████████████████| 71kB 8.5MB/s 
     |████████████████████████████████| 102kB 11.6MB/s 
  Created wheel for googletrans: filename=googletrans-3.0.0-cp36-none-any.whl size=15736 sha256=41b7b3122e33c9df3342cbb9a16181b61a416c00dc8bba708fa73a4225c295bc
  Stored in directory: /root/.cache/pip/wheels/28/1a/a7/eaf4d7a3417a0c65796c547cff4deb6d79c7d14c2abd29273e
  Created wheel for context

In [2]:
import numpy as np                                                 # Implemennts milti-dimensional array and matrices
from random import randint
import pandas as pd                                                # For data manipulation and analysis
import mysql.connector                                             # Pacakage for mysql connection
from mysql.connector import Error
# Matplotlib and associated plotting modules
import matplotlib.pyplot as plt                                    # Plotting library for Python programming language and it's numerical mathematics extension NumPy
import matplotlib.cm as cm
import matplotlib.colors as colors
import seaborn as sns                                              # Provides a high level interface for drawing attractive and informative statistical graphics
%matplotlib inline
sns.set()

from geopy.geocoders import Nominatim # convert an address into latitude and longitude values
import folium # map rendering library

from IPython.display import display
pd.set_option('display.max_columns', 100)
pd.set_option('display.max_info_columns',100)
pd.set_option('display.max_rows', 100)

import warnings
warnings.filterwarnings("ignore", category=FutureWarning)
warnings.filterwarnings("ignore", category=DeprecationWarning)
warnings.filterwarnings("ignore", category=FutureWarning)
from subprocess import check_output

/usr/local/lib/python3.6/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


**Loading data**

In [3]:
# We will read Events data from the Google Drive | Other 2 datasets will be read directly from the MySQL datasource provided by INSAID
from google.colab import drive
drive.mount('/content/drive')

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&response_type=code&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly

Enter your authorization code:
··········
Mounted at /content/drive


In [4]:
# defining constants to be used through-out the book
AndhraPradesh = 'AndhraPradesh'
Pondicherry = 'Pondicherry'
Mizoram = 'Mizoram'
AndamanandNicobarIslands = 'AndamanandNicobarIslands'
Meghalaya = 'Meghalaya'
HimachalPradesh = 'HimachalPradesh'
Karnataka = 'Karnataka'
states = [AndhraPradesh, Pondicherry, Mizoram, AndamanandNicobarIslands, Meghalaya, HimachalPradesh, Karnataka]

In [5]:
# Function to connect to MySQL datasource & reading a table into dataframe
def get_datatable(tableName): 
    try:
        connection = mysql.connector.connect(host='cpanel.insaid.co',
                                         port='3306',
                                         database='Capstone1',
                                         user='student',
                                         password='student')
        if connection.is_connected():
            sql = "SELECT * FROM " + tableName
            df = pd.read_sql(sql, con=connection)
    except Error as e:
        print("Error while connecting to MySQL", e)
    finally:
        if (connection.is_connected()):
            connection.close()
            print("MySQL connection is closed")
    return df

In [6]:
# Read all 3 datasets into respective dataframes

# Importing events data from Google Drive into DataFrame
dfevents = pd.read_csv("/content/drive/My Drive/events_data.csv",dtype=str)

# reading data from mysql gender_age_train
dfdemographic = get_datatable("gender_age_train")

# reading data from mysql phone_brand_device_model
dfphone = get_datatable("phone_brand_device_model")

MySQL connection is closed
MySQL connection is closed


**Analyze columns and their datatypes for each dataframes**

In [ ]:
dfevents.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3252950 entries, 0 to 3252949
Data columns (total 7 columns):
 #   Column     Dtype 
---  ------     ----- 
 0   event_id   object
 1   device_id  object
 2   timestamp  object
 3   longitude  object
 4   latitude   object
 5   city       object
 6   state      object
dtypes: object(7)
memory usage: 173.7+ MB


In [ ]:
dfdemographic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 74645 entries, 0 to 74644
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   device_id  74645 non-null  int64 
 1   gender     74645 non-null  object
 2   age        74645 non-null  int64 
 3   group      74645 non-null  object
dtypes: int64(2), object(2)
memory usage: 2.3+ MB


In [ ]:
dfphone.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 87726 entries, 0 to 87725
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   device_id     87726 non-null  int64 
 1   phone_brand   87726 non-null  object
 2   device_model  87726 non-null  object
dtypes: int64(1), object(2)
memory usage: 2.0+ MB


***Obseravations from Info***



*   device_id field is a common field in all 3 tables.
*   device_id field is **float** data type in ***dfevents*** and **int** in dfdemographic and dfphone data. 
*   We will need to make device_id in ***dfevents*** to **int** and then **string(object)**.
*   Similarly we will convert device_id from ***dfdemographic*** and ***dfphone*** to **string(object)**




**Further having look at the sample data**

In [ ]:
dfevents.head()

event_id             device_id            timestamp          longitude  \
0  2765368   2973347786994914304  2016-05-07 22:52:05  77.22567599999998   
1  2955066   4734221357723754978  2016-05-01 20:44:16          88.388361   
2   605968  -3264499652692492784  2016-05-02 14:23:04  77.25680899999998   
3   448114   5731369272434022440  2016-05-03 13:21:16  80.34361299999998   
4   665740    338888002570799963  2016-05-06 03:51:05          85.997745   

             latitude      city       state  
0  28.730140000000002     Delhi       Delhi  
1           22.660325  Calcutta  WestBengal  
2           28.757906     Delhi       Delhi  
3           13.153332   Chennai   TamilNadu  
4           23.842609    Bokaro   Jharkhand

In [ ]:
dfdemographic.head()

device_id gender  age   group
0 -8076087639492063270      M   35  M32-38
1 -2897161552818060146      M   35  M32-38
2 -8260683887967679142      M   35  M32-38
3 -4938849341048082022      M   30  M29-31
4   245133531816851882      M   30  M29-31

In [ ]:
dfphone.head()

device_id phone_brand device_model
0  1877775838486905855        vivo          Y13
1 -3766087376657242966          小米         V183
2 -6238937574958215831        OPPO          R7s
3  8973197758510677470          三星        A368t
4 -2015528097870762664          小米      红米Note2

**Observations**
*  Some device_id's are having ***negative*** signs.
*  phone_brand in dfphone is having some ***chinese values*** which need to be replaced to english for better data analysis and understanding.

<a id=section6></a>
### 6. Data Pre-Processing

**Checking for missing values and , issues and treatment for dfevents**


---


In [7]:
total = dfevents.isnull().sum().sort_values(ascending=False)
percent = ((dfevents.isnull().sum()/dfevents.isnull().count())*100).sort_values(ascending=False)
missing_data = pd.concat([total, percent], axis=1, keys=['Total', 'Percent'])
missing_data.head(10)

Total   Percent
device_id    453  0.013926
latitude     423  0.013004
longitude    423  0.013004
state        377  0.011589
city           0  0.000000
timestamp      0  0.000000
event_id       0  0.000000

***Observations***

*   State has 423 missing values, we can fix on the basis of city.
*   Device _id , latitude & logitude missinng value will be treated separatley while analyzing with folium 






Treating missing states

In [8]:
#function for treating missing state values
def replace_missing_state(df):
    missing_state_city = df[df['state'].isnull()]['city'].unique()
    missing_state = []
    for city in missing_state_city:
        missing_state.append(df.loc[df['city'] == city]['state'].unique()[0])
    for i in range(0, len(missing_state)):
        cond = (df['city'] == missing_state_city[i])
        df.loc[cond, 'state'] = df.loc[cond, 'state'].fillna(missing_state[i])
    print('Missing state count: ' + str(df.state.isnull().sum()))

In [9]:
replace_missing_state(dfevents)

Missing state count: 0


In [10]:
# Check the percentage of other null values Remaining.
total = dfevents.isnull().sum().sort_values(ascending=False)
percent = ((dfevents.isnull().sum()/dfevents.isnull().count())*100).sort_values(ascending=False)
missing_data = pd.concat([total, percent], axis=1, keys=['Total', 'Percent'])
missing_data.head(10)

Total   Percent
device_id    453  0.013926
latitude     423  0.013004
longitude    423  0.013004
state          0  0.000000
city           0  0.000000
timestamp      0  0.000000
event_id       0  0.000000

## Feature Extraction
### Extract Year, Month, Hour, Weekday from Timestamp

In [11]:
##code to extract year, month, hour, weekday from time stamp
dfevents['timestamp']= pd.to_datetime(dfevents['timestamp']) 
dfevents['year']= dfevents.timestamp.dt.year
dfevents['month']= dfevents.timestamp.dt.month
dfevents['day']= dfevents.timestamp.dt.day
dfevents['hour']= dfevents.timestamp.dt.hour 

In [ ]:
dfevents.head(20)

event_id             device_id           timestamp          longitude  \
0   2765368   2973347786994914304 2016-05-07 22:52:05  77.22567599999998   
1   2955066   4734221357723754978 2016-05-01 20:44:16          88.388361   
2    605968  -3264499652692492784 2016-05-02 14:23:04  77.25680899999998   
3    448114   5731369272434022440 2016-05-03 13:21:16  80.34361299999998   
4    665740    338888002570799963 2016-05-06 03:51:05          85.997745   
5   1078723   -512424213979236423 2016-05-02 02:21:20  83.39824399999998   
6   2948755   5536513450525271144 2016-05-03 16:20:14          80.359796   
7    280014  -8879643774076165575 2016-05-05 13:06:01          78.155397   
8   2309828   3020769016382465619 2016-05-04 09:24:50          77.263922   
9    720049   4928555839839669963 2016-05-06 19:00:52  72.84479499999998   
10  3219828  -3433061233631405742 2016-05-03 00:27:21          80.290779   
11    13998   3978025911571213813 2016-05-06 13:12:42          73.875196   
12  2334601   -601883305299256779 2016-05-05 11:17:48          83.380111   
13  2633127  -3090166955280575327 2016-05-05 04:42:59  77.28591999999998   
14   460645   7983871429718484702 2016-05-04 21:04:40  80.33543499999998   
15   998037   7594215820567965078 2016-05-03 20:59:09          80.306871   
16  2906758  -2168523477348251699 2016-05-03 10:58:25          80.344605   
17  2224400   7566051280927280645 2016-05-01 00:07:47          77.240478   
18  2045001  -3656132883647868884 2016-05-01 07:35:59          77.251554   
19  2072637   3644267756307556227 2016-05-04 05:02:28          80.327388   

              latitude           city          state  year  month  day  hour  
0   28.730140000000002          Delhi          Delhi  2016      5    7    22  
1            22.660325       Calcutta     WestBengal  2016      5    1    20  
2            28.757906          Delhi          Delhi  2016      5    2    14  
3            13.153332        Chennai      TamilNadu  2016      5    3    13  
4            23.842609         Bokaro      Jharkhand  2016      5    6     3  
5   17.768148999999994  Visakhapatnam  AndhraPradesh  2016      5    2     2  
6            13.128174        Chennai      TamilNadu  2016      5    3    16  
7   16.390327000000006       Wanparti  AndhraPradesh  2016      5    5    13  
8   28.757914000000003          Delhi          Delhi  2016      5    4     9  
9            18.974929         Mumbai    Maharashtra  2016      5    6    19  
10           13.115427        Chennai      TamilNadu  2016      5    3     0  
11           18.615791           Pune    Maharashtra  2016      5    6    13  
12  17.828583000000005  Visakhapatnam  AndhraPradesh  2016      5    5    11  
13           28.758381          Delhi          Delhi  2016      5    5     4  
14           13.189053        Chennai      TamilNadu  2016      5    4    21  
15            13.17297        Chennai      TamilNadu  2016      5    3    20  
16           13.169308        Chennai      TamilNadu  2016      5    3    10  
17           28.715664          Delhi          Delhi  2016      5    1     0  
18  28.737959000000004          Delhi          Delhi  2016      5    1     7  
19           13.179729        Chennai      TamilNadu  2016      5    4     5

In [ ]:
dfevents.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3252950 entries, 0 to 3252949
Data columns (total 11 columns):
 #   Column     Dtype         
---  ------     -----         
 0   event_id   object        
 1   device_id  object        
 2   timestamp  datetime64[ns]
 3   longitude  object        
 4   latitude   object        
 5   city       object        
 6   state      object        
 7   year       int64         
 8   month      int64         
 9   day        int64         
 10  hour       int64         
dtypes: datetime64[ns](1), int64(4), object(6)
memory usage: 273.0+ MB


Checking length of device_id values

In [12]:
dfevent_deviceidlen = dfevents.loc[:,['device_id']]
dfevent_deviceidlen['length_of_device_id'] = dfevent_deviceidlen['device_id'].str.len()
dfevent_deviceidlen = dfevent_deviceidlen.groupby(by=['length_of_device_id']).count()
dfevent_deviceidlen.rename(columns={'device_id': 'number_of_devices'}, inplace=True)
print('Length of device_id in dfevent table')
dfevent_deviceidlen.sort_values(by= ['number_of_devices'], ascending=False)

Length of device_id in dfevent table


number_of_devices
length_of_device_id                   
19.0                           1626271
20.0                           1427959
18.0                            178481
17.0                             18407
16.0                              1289
15.0                                89
14.0                                 1

**Observations**
*    Majority of values are between 17-19 characters largest having count of 2901640

Filtering out data for the states in scope of the requirement.

In [13]:
# Code to filter out state specific rows
dffiltered_events= dfevents[dfevents.state.isin(states)]
dffiltered_events.shape

(428244, 11)

In [ ]:
dffiltered_events.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 428244 entries, 5 to 3252948
Data columns (total 11 columns):
 #   Column     Non-Null Count   Dtype         
---  ------     --------------   -----         
 0   event_id   428244 non-null  object        
 1   device_id  428175 non-null  object        
 2   timestamp  428244 non-null  datetime64[ns]
 3   longitude  428181 non-null  object        
 4   latitude   428181 non-null  object        
 5   city       428244 non-null  object        
 6   state      428244 non-null  object        
 7   year       428244 non-null  int64         
 8   month      428244 non-null  int64         
 9   day        428244 non-null  int64         
 10  hour       428244 non-null  int64         
dtypes: datetime64[ns](1), int64(4), object(6)
memory usage: 39.2+ MB


**Checking for missing values,issues and treatment for dfdemographic**

---




In [14]:
total = dfdemographic.isnull().sum().sort_values(ascending=False)
percent = ((dfdemographic.isnull().sum()/dfdemographic.isnull().count())*100).sort_values(ascending=False)
missing_data = pd.concat([total, percent], axis=1, keys=['Total', 'Percent'])
missing_data.head(20)

Total  Percent
group          0      0.0
age            0      0.0
gender         0      0.0
device_id      0      0.0

It seems there is no missing data. Lets check if there is any problem with age by using describe.

In [ ]:
dfdemographic.describe(include='all')

device_id gender           age   group
count   7.464500e+04  74645  74645.000000   74645
unique           NaN      2           NaN      12
top              NaN      M           NaN  M23-26
freq             NaN  47904           NaN    9605
mean   -7.491354e+14    NaN     31.410342     NaN
std     5.327150e+18    NaN      9.868735     NaN
min    -9.223067e+18    NaN      1.000000     NaN
25%    -4.617367e+18    NaN     25.000000     NaN
50%    -1.841362e+16    NaN     29.000000     NaN
75%     4.636656e+18    NaN     36.000000     NaN
max     9.222849e+18    NaN     96.000000     NaN

In [15]:
dfdemographic['gender'].unique()

array(['M', 'F'], dtype=object)

In [ ]:
dfdemographic[dfdemographic['age'] < 6]

device_id gender  age group
10385  3553057874282315257      M    1  M22-

In [ ]:
dfdemographic['group'].unique()

array(['M32-38', 'M29-31', 'F24-26', 'F33-42', 'F27-28', 'M39+', 'M23-26',
       'M27-28', 'M22-', 'F43+', 'F23-', 'F29-32'], dtype=object)

**Observations**


*   There are no missing values for dfdemographic
*   Age seems to have problem since the minimum value is 1, a kid of one year might not be using mobile, further brainstorming required on same, only one such record is found.
*   Gender field has no issue.
*   There seems to be problem with group field , it looks inconsistent(M22-,F23-,F43+,M39+) . We will need to study further,fix the same or create new grouping.



Fixing data type for device_id and replacing '-' in same

In [16]:
##dfdemographic['device_id'] = dfdemographic['device_id'].str.replace('-', '')
dfdemographic['device_id'] = dfdemographic['device_id'].astype(str)

Checking length of device_id values

In [17]:
dfdemographic_deviceidlen = dfdemographic.loc[:,['device_id']]
dfdemographic_deviceidlen['length_of_device_id'] = dfdemographic_deviceidlen['device_id'].str.len()
dfdemographic_deviceidlen = dfdemographic_deviceidlen.groupby(by=['length_of_device_id']).count()
dfdemographic_deviceidlen.rename(columns={'device_id': 'number_of_devices'}, inplace=True)
print('Length of device_id in dfdemographic table')
dfdemographic_deviceidlen.sort_values(by= ['number_of_devices'], ascending=False)

Length of device_id in dfdemographic table


number_of_devices
length_of_device_id                   
19                               36961
20                               33317
18                                3915
17                                 409
16                                  39
15                                   3
14                                   1

**Observation**
*    Majority of values are between 17-19 characters largest having count of 66591

**Checking for missing values,issues and treatment for dfphone**

---


In [ ]:
total = dfphone.isnull().sum().sort_values(ascending=False)
percent = ((dfphone.isnull().sum()/dfphone.isnull().count())*100).sort_values(ascending=False)
missing_data = pd.concat([total, percent], axis=1, keys=['Total', 'Percent'])
missing_data.head(20)

Total  Percent
device_model      0      0.0
phone_brand       0      0.0
device_id         0      0.0

**Observations**
* No Missing values in df phone 

Fixing data type for device_id and replacing '-' in same

In [18]:
##dfphone['device_id'] = dfphone['device_id'].str.replace('-', '')
dfphone['device_id'] = dfphone['device_id'].astype(str)

Checking length of device_id values

In [ ]:
dfphone_deviceidlen = dfphone.loc[:,['device_id']]
dfphone_deviceidlen['length_of_device_id'] = dfphone_deviceidlen['device_id'].str.len()
dfphone_deviceidlen = dfphone_deviceidlen.groupby(by=['length_of_device_id']).count()
dfphone_deviceidlen.rename(columns={'device_id': 'number_of_devices'}, inplace=True)
print('Length of device_id in dfphone table')
dfphone_deviceidlen.sort_values(by= ['number_of_devices'], ascending=False)

Length of device_id in dfphone table


number_of_devices
length_of_device_id                   
19                               43401
20                               39206
18                                4600
17                                 468
16                                  47
15                                   3
14                                   1

**Observations**
*    Majority of values are between 17-19 characters largest having count of 78254

Treatment for chinese language in brand name column.

In [ ]:
dfphone['phone_brand'].unique()

array(['vivo', '小米', 'OPPO', '三星', '酷派', '联想 ', '华为', '奇酷', '魅族', '斐讯',
       '中国移动', 'HTC', '天语', '至尊宝', 'LG', '欧博信', '优米', 'ZUK', '努比亚', '惠普',
       '尼比鲁', '美图', '乡米', '摩托罗拉', '梦米', '锤子', '富可视', '乐视', '海信', '百立丰',
       '一加', '语信', '海尔', '酷比', '纽曼', '波导', '朵唯', '聆韵', 'TCL', '酷珀', '爱派尔',
       'LOGO', '青葱', '果米', '华硕', '昂达', '艾优尼', '康佳', '优购', '邦华', '赛博宇华',
       '黑米', 'Lovme', '先锋', 'E派', '神舟', '诺基亚', '普耐尔', '糖葫芦', '亿通', '欧新',
       '米奇', '酷比魔方', '蓝魔', '小杨树', '贝尔丰', '糯米', '米歌', 'E人E本', '西米', '大Q',
       '台电', '飞利浦', '唯米', '大显', '长虹', '维图', '青橙', '本为', '虾米', '夏新', '帷幄',
       '百加', 'SUGAR', '欧奇', '世纪星', '智镁', '欧比', '基伍', '飞秒', '德赛', '易派',
       '谷歌', '金星数码', '广信', '诺亚信', 'MIL', '白米', '大可乐', '宝捷讯', '优语', '首云',
       '瑞米', '瑞高', '沃普丰', '摩乐', '鲜米', '凯利通', '唯比', '欧沃', '丰米', '恒宇丰',
       '奥克斯', '西门子', '欧乐迪', 'PPTV'], dtype=object)

In [19]:
! pip install google translator

In [20]:
# Translating brands in English for better understanding and standardization
import googletrans
from googletrans import Translator
translator = Translator()
translations = {}
for column in dfphone.columns[1:2]:
    # unique elements of the column
    print('processing {} column'.format(column))
    unique_elements = dfphone[column].unique()
    print (len(unique_elements))
    for element in unique_elements:
        # add translation to the dictionary
        translations[element] = translator.translate(element).text
dfphone.replace(translations,inplace=True)

processing phone_brand column
116


In [ ]:
dfphone['phone_brand'].unique()

array(['vivo', 'Xiaomi', 'OPPO', 'Samsung', 'Coolpad', 'Lenovo', 'Huawei',
       'Qiku', 'Meizu', 'Feixun', 'China Mobile', 'HTC', 'Amagatarai',
       'Extreme treasure', 'LG', 'Obersin', 'Quality rice', 'YOU',
       'Nubian', 'HP', 'Nibiru', 'Mito', 'Xiangmi', 'Motorola', 'Meng Mi',
       'hammer', 'InFocus', 'Letv', 'Hisense', 'Bilifeng', 'One plus',
       'Verbal letter', 'Haier', 'Cool', 'Newman', 'waveguide', 'Dove',
       'Listen to rhyme', 'TCL', 'Amber', 'Eppel', 'LOGO', 'Shallots',
       'Guomi', 'ASUS', 'Onda', 'Ayuni', 'Konka', 'Ugo', 'Banghua',
       'Cyber \u200b\u200bYuhua', 'black rice', 'Love Me', 'pioneer',
       'E Pie', 'Shenzhou', 'Nokia', 'Pnel', 'Candied', 'Toru', 'Ou Xin',
       'Mickey', "Cube Rubik's Cube", 'Blue demon', 'Little Poplar',
       'Belfont', 'Glutinous rice', 'Mige', 'E people E Ben', 'Simi',
       'Big Q', 'Taipower', 'Philips', 'Vimi', 'Daxian', 'Changhong',
       'Vito', 'Green orange', 'Originally', 'Shrimp', 'Xia Xin',
       'Cur

**Observations**
*  Using GoogleTrans API, we could translate the brands name. However, we cannot assure the authenticity of the translation whether it is correct. A quick look at a small sample suggests that the translation went per the expectation. 
*   As there are many phone brands available globally, it is difficult to cross-check whether such a brand indeed exist. We will rely on GoogleTrans' o/p & proceed further.
*   Also, we are skipping the translation of 1467 unique device models. We do not see it is important for our analysis.


**Merging Dataframes into one Dataframe for detailed analysis**

In [21]:
dffiltered_events.shape

(428244, 11)

In [22]:
dfdemographic.shape

(74645, 4)

In [23]:
dfevents_demographic = pd.merge(dffiltered_events, dfdemographic, how='inner', on=['device_id'])

In [24]:
dfevents_demographic.shape

(428175, 14)

In [25]:
dfphone.shape

(87726, 3)

In [26]:
dfevents_combined = pd.merge(dfevents_demographic, dfphone, how='inner', on=['device_id']) 

In [33]:
# Code added by Raju on July 18th
# Write the DataFrame to CSV file.
with open('/content/drive/My Drive/dashboard.csv', 'w') as f:
  dfevents_combined.to_csv(f)

In [27]:
dfevents_combined.shape

(428175, 16)

In [28]:
dfevents_combined.device_id.nunique()

11704

In [ ]:
total = dfevents_combined.isnull().sum().sort_values(ascending=False)
percent = ((dfevents_combined.isnull().sum()/dfevents_combined.isnull().count())*100).sort_values(ascending=False)
missing_data = pd.concat([total, percent], axis=1, keys=['Total', 'Percent'])
missing_data.head(20)

Total   Percent
latitude         63  0.014714
longitude        63  0.014714
device_model      0  0.000000
phone_brand       0  0.000000
group             0  0.000000
age               0  0.000000
gender            0  0.000000
hour              0  0.000000
day               0  0.000000
month             0  0.000000
year              0  0.000000
state             0  0.000000
city              0  0.000000
timestamp         0  0.000000
device_id         0  0.000000
event_id          0  0.000000

**There are 63 records having null latitude & longitude will check while ploting & repalce missing values**

In [ ]:
dfgeodata = dfevents_combined[['city','state','latitude','longitude']]
dfgeodata = dfgeodata.drop_duplicates(['city','state','latitude','longitude'])
dfgeodata.shape

(11709, 4)

### **Identify incorrect mapping of latitude and longitude for each state in study**

Checking for AndhraPradesh

In [ ]:
dfgeodataAP = dfgeodata[dfgeodata['state'] == AndhraPradesh]
dfgeodataAP.isnull().sum()
dfgeodataAP.dropna(inplace=True)
dfgeodataAP.isnull().sum()

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


city         0
state        0
latitude     0
longitude    0
dtype: int64

In [ ]:
address = 'Guntur, IN'

geolocator = Nominatim(user_agent="to_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of {} are {}, {}.'.format(address, latitude, longitude))

The geograpical coordinate of Guntur, IN are 16.2915189, 80.4541588.


In [ ]:
# create map of Toronto using latitude and longitude values
map_ap = folium.Map(location=[latitude, longitude], zoom_start=2,tiles='Stamen Terrain')

# add markers to map
for lat, lng, state, city in zip(dfgeodataAP['latitude'], dfgeodataAP['longitude'], dfgeodataAP['state'], dfgeodataAP['city']):
    label = '{}, {},Latitude:{},Longitude:{}'.format(state, city,lat,lng)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_ap)  
    
map_ap

RecursionError: ignored

## Code for statewise- city wise maps

In [ ]:
#function for adding markers on map
def addmap_marker(dfmarkCity,selcolor,objmap):
    # add markers to map
    for lat, lng, state, city in zip(dfmarkCity['latitude'], dfmarkCity['longitude'], dfmarkCity['state'], dfmarkCity['city']):
        label = '{}, {},Latitude:{},Longitude:{}'.format(state, city,lat,lng)
        label = folium.Popup(label, parse_html=True)
        folium.CircleMarker(
            [lat, lng],
            radius=5,
            popup=label,
            color=selcolor,
            fill=True,
            fill_color=selcolor,
            fill_opacity=0.7,
            parse_html=False).add_to(objmap)  
    return

In [ ]:
#function for plotting map for selected state
def plot_map(state): 
    latitude = 20.5937
    longitude = 78.9629
    map_india = folium.Map(location=[latitude, longitude], zoom_start=5,tiles='Stamen Terrain')
    dfgeodata = dfevents_combined[['city','state','latitude','longitude']]
    dfgeodata.dropna(inplace=True)
    dfgeodata = dfgeodata[dfgeodata['state']==state].drop_duplicates(['city','state','latitude','longitude'])
    dfgeodataUCities = pd.DataFrame(dfgeodata['city'].unique(),columns=['city'])

    dfgeodataUCities["color"] = ""
    colors = []
    for i in range(len(dfgeodataUCities)):
          colors.append('#%06X' % randint(0, 0xFFFFFF))
    #print(colors)        
    
    dfgeodataUCities["color"] = colors
    
    for i in range(len(dfgeodataUCities)):
            dfgeodataSelCity = dfgeodata[dfgeodata['city']==dfgeodataUCities['city'][i]]
            selcolor = dfgeodataUCities['color'][i]
            addmap_marker(dfgeodataSelCity,selcolor,map_india)
    return map_india

In [ ]:
# Map for AndraPradesh  
plot_map(AndhraPradesh)

**9 records in AndhraPradesh have incorrect coordinates** <br>
Below bad coordinates have 3 records each which need to be replaced based on their corresponding city-state values 
Latitude:34.5553,Longitude:69.2075 -- Kabul <br>
Latitude:25.2048,Longitude:55.2708 -- Dubai <br>
Latitude:41.8719,Longitude:12.5674 -- Rome

In [ ]:
# Map for Karnataka
plot_map(Karnataka)

**3 records in Karnataka have incorrect coordinates** <br>
There are 3 records having below bad coordinates which need to be replaced based on their corresponding city-state values 
Latitude:25.2048,Longitude:55.270 -- Dubai

In [ ]:
# Map for Pondicherry
plot_map(Pondicherry)

In [ ]:
# Map for Mizoram
plot_map(Mizoram)

In [ ]:
# Map for AndamanandNicobarIslands
plot_map(AndamanandNicobarIslands)

In [ ]:
# Map for Meghalaya
plot_map(Meghalaya)

In [ ]:
# Map for HimachalPradesh
plot_map(HimachalPradesh)

**Final observations on latitude longitude**<br>
1. Using Folium, bad coordinates have been identified for the state of AndhraPradesh & Karnataka which need to be replaced with the right ones <br>
2. There are 63 reccords having null latitude-longitude. All of these records are in the state of AndhraPradesh. These need to be replaced with the right ones <br>

**Below function takes care of above 2 issues.**

In [ ]:
# logic to replace the lat-long outliers detected | replace the "bad" coordinates with the most frequent coordinate from which the devices are using the INSAID Telecom service

def FindAndReplaceCoordinate(state, latitude, longitude):
  
  """Objective of this function is to replace the bad coordinates with the right ones.
     Logic:
     1. get all the records for the passed state, latitude, longitude -- these records have bad coordinates
     2. iterate over all the records and do the following:
        2.1 extract the city & state information of that record
        2.2 get the most frequent coordinates for city & state pair using GetMostFrequentCoordinate() function
        2.3 replace the bad coordinates with the right ones -- using ReplaceBadCoordinates() function
  """
  # Get all the records having passed latitude and longitude
  if (latitude is None) | (longitude is None): 
    # case: handle missing lat-long values | get all the records having missing lat-long values
    print('*'*5 + 'Handling missing lat-long scenario'+5*'*')
    dfbadcoordinates = dfevents_combined[dfevents_combined['state'] == AndhraPradesh] # select the subset of the dataset
    dfbadcoordinates = dfbadcoordinates.loc[np.array(dfbadcoordinates.isnull().loc[:, ['latitude']]), ['city', 'state']] # find all the records having null latitude (or longitude)
  else: 
    # for other scenarios such as outliers
    print('*'*5+'Handling outliers/mismatch scenario'+5*'*')
    dfbadcoordinates  = dfevents_combined.loc[(dfevents_combined['state'] == state) & (dfevents_combined['latitude'] == latitude) & (dfevents_combined['longitude'] == longitude), ['city', 'state']]

  print('Found {} records for which lat and long are to be replaced'.format(len(dfbadcoordinates)))
  # Iterate over all the records found for these bad coordinates & replace with right coordinates for each record based on their city & state
  for record in dfbadcoordinates.iterrows(): #returns a tuple
    
    record_index = record[0] #extract the index of the record
    city, state = record[1][0], record[1][1] #extract city & state from the tuple
    
    # Get new coordinates based on most frequent coordinates in city & state pair
    df_frequent_coordinates = GetMostFrequentCoordinate(city, state)
    most_frequent_latitude = df_frequent_coordinates['latitude'].iloc[0]
    most_frequent_longitude = df_frequent_coordinates['longitude'].iat[0]

    print('Replacing coords for the record id {}, state: {}, outlier_latitude: {}, outlier_longitude: {}, new_latitude: {}, new_longitude: {}'.format(record_index, state, latitude, longitude, most_frequent_latitude, most_frequent_longitude))
    
    # Replace the bad coordinates with the new ones
    ReplaceBadCoordinates(record_index, most_frequent_latitude, most_frequent_longitude)


# function to find the most frequent coordinates based on city & state pair -> this will be the new coordinates for the outlier coordinates
def GetMostFrequentCoordinate(city, state):
  return dfevents_combined.loc[(dfevents_combined['city'] == city) & (dfevents_combined['state'] == state), ['device_id', 'latitude', 'longitude']] \
                 .groupby(by=['latitude', 'longitude']).count().reset_index() \
                 .sort_values(by=['device_id'], ascending=False) \
                 .head(1) \
                 .loc[:, ['latitude', 'longitude']]

# function to replace the outlier coordinates with the new ones
def ReplaceBadCoordinates(record_index, new_latitude, new_longitude):
  dfevents_combined.loc[[record_index], ['latitude', 'longitude']] = [new_latitude, new_longitude]
 

                 
# Prepare the list of states having outlier coordinates which need to be replaced
dict_outlier_coordinates = {AndhraPradesh:[{'latitude': 34.5553, 'longitude': 69.2075, 'mappedto': 'Kabul'},\
                                           {'latitude': 25.2048, 'longitude': 55.2708, 'mappedto': 'Dubai'},\
                                           {'latitude': 41.8719, 'longitude': 12.5674, 'mappedto': 'Rome'}, \
                                           {'latitude': None, 'longitude': None, 'mappedto': 'Unmapped'}], \
                            Karnataka:[{'latitude': 25.2048, 'longitude': 55.2708, 'mappedto': 'Dubai'}]
                            }


# 4. Drilling time - replace the outliers
for state in dict_outlier_coordinates:
  for coordinates in dict_outlier_coordinates[state]:
    
    # Outliers to be replaced
    outlier_latitude = coordinates['latitude']
    outlier_longitude = coordinates['longitude']

    FindAndReplaceCoordinate(state, outlier_latitude, outlier_longitude)    

In [ ]:
# confirming that the null records have been fixed
dfevents_combined.isnull().sum()

In [ ]:
# Code for new_age_group 
# list of our conditions
conditions = [
    (dfevents_combined['age'] <= 20),
    (dfevents_combined['age'] > 20) & (dfevents_combined['age'] <= 30),
    (dfevents_combined['age'] > 30) & (dfevents_combined['age'] <= 40),
    (dfevents_combined['age'] > 40)
    ]

# list of the values we want to assign for each condition
values = ['below 20', '21 to 30', '31 to 40', 'above 40']

# new column and use np.select to assign values to it using our lists as arguments
dfevents_combined['new_group'] = np.select(conditions, values)

# display updated DataFrame
dfevents_combined.head()

In [ ]:
# Code for dataframe with unique users
dfunique_users = dfevents_combined[['device_id','state','gender','age','group','phone_brand','new_group']]

In [ ]:
dfunique_users.drop_duplicates(inplace=True)

#Code below if requred for further grouping
#dfuuser = dfunique_users.groupby(by=['state'], as_index=False).agg({'device_id': pd.Series.nunique})
#dfuuser.rename(columns={'device_id': 'userscount'}, inplace=True)

In [ ]:
dfunique_users.shape

# <h2 style="color:red">Data Visualization</h2>

### Point 1: Distribution of Users(device_id) across States.

In [ ]:
# code to see distribution of User's across states
dfevents_combined['state'].value_counts().plot(kind = 'bar',figsize=(15,7), colormap='Dark2', fontsize=13)


In [ ]:
ax = sns.catplot(y='state',kind='count',height=12,order=dfevents_combined.state.value_counts().index,data=dfevents_combined)

In [ ]:
dfevents_combined['state'].value_counts()

### Observation for Point 1:
> - Majority of the Users are in Andhrapradesh state. Very few users are in Himachalpradesh, AndamanandNicobarIslands, Meghalaya ,Mizoram  states                                     

## 2 . Distribution of Users across Phone Brands(Consider only 10 Most used Phone Brands) - Unique users


In [ ]:
dfbrand=dfevents_combined.groupby(['phone_brand'])['device_id'].nunique()
dfbrand.sort_values(inplace=True, ascending=False)
# Plotting the Count and Proportional Distribution of brands based on Users
plt.figure(figsize=(16, 8))

plt.subplot(1, 2, 1)

# Plotting the count of Brand
dfbrand.head(10).plot(kind='bar', colormap='Dark2', fontsize=12)
plt.xlabel('')
plt.ylabel('Number of users', fontsize=16)
plt.title('Distribution of users across top 10 Phone Brands', fontsize=18)

plt.subplot(1, 2, 2)

# Plotting Proportional Distribution of top 10 Brands
dfbrand.head(10).plot(kind='pie', autopct='%1.1f%%', pctdistance=0.6,  radius=0.9, 
                                                   colormap='viridis', legend=True, wedgeprops={'edgecolor':'black', 'linewidth':2}, 
                                                   textprops={'fontsize':20, 'color':'white'})
plt.ylabel('Brand', fontsize=16)
plt.title('Proportional Distribution of top 10 Brands', fontsize=18)

**Observation 2:**
# In top 10 brands, 67% are contributed by 3 brands Xiaomi, Samsung & Huawei.

In [ ]:
#new_age_hroup category count
dfevents_combined.new_group.value_counts()

In [ ]:
#Distribution of Users across Gender.

genders = dfevents_combined.gender.value_counts()
print(genders);

# only "explode" the 2nd slice
explode = (0, 0.1) 

plt.pie(genders.values,explode=explode,labels=genders.keys(),autopct='%1.1f%%',
        shadow=True, startangle=90)
plt.axis('equal')
plt.show()

In [ ]:
#Distribution of Users across Age Segments.

groups = dfevents_combined.new_group.value_counts()

print(groups);

explode = (0.05,0.05,0.05,0.05)  

#add colors
colors = ['#ff9999','#66b3ff','#99ff99','#ffcc99']

 
fig1, ax1 = plt.subplots(figsize = (12,8))

ax1.pie(groups.values, explode=explode, colors=colors, labels=groups.keys(),autopct='%1.1f%%',
        shadow=True, startangle=90)

#draw circle
centre_circle = plt.Circle((0,0),0.70,fc='white')
fig = plt.gcf()
fig.gca().add_artist(centre_circle)

plt.axis('equal')
plt.show()

# 5 & 7. Distribution of Phone Brands(Consider only 10 Most used Phone Brands) for each Age Segment, State, Gender.


In [ ]:
from itertools import cycle, islice
top10_brands = dfunique_users.phone_brand.value_counts().index[:10]
df_events_combined_top10_brands_usage = dfunique_users[dfunique_users.phone_brand.isin(top10_brands)]

df = df_events_combined_top10_brands_usage.groupby('state').apply(lambda x: x['gender'].value_counts())
df = df.unstack().fillna(0)
df.plot.bar(stacked=True,figsize=(18,5),width=.2, colormap='cividis')

df = df_events_combined_top10_brands_usage.groupby('phone_brand').apply(lambda x: x['new_group'].value_counts())
df = df.unstack().fillna(0)
df.plot.bar(stacked=True,figsize=(18,5),width=.2, colormap='gist_heat')


df = df_events_combined_top10_brands_usage.groupby('phone_brand').apply(lambda x: x['state'].value_counts())
df = df.unstack().fillna(0)
df.plot.bar(stacked=True,figsize=(18,5),width=.2, colormap='autumn_r')

df = df_events_combined_top10_brands_usage.groupby('gender').apply(lambda x: x['phone_brand'].value_counts())
df = df.unstack().fillna(0)
df.plot.bar(stacked=True,figsize=(18,5),width=.05, colormap='magma_r')


df = df_events_combined_top10_brands_usage.groupby('state').apply(lambda x: x['new_group'].value_counts())
df = df.unstack().fillna(0)
df.plot.bar(stacked=True,figsize=(18,5),width=.2, colormap='Paired')

> -  Distribution of Phone Brands across states

In [ ]:
# Check how the top 10 phone brands are distributed across states
df = dfevents_combined.groupby(['phone_brand','state'])['state']

topbrands = dfevents_combined[dfevents_combined.state. isin(['AndhraPradesh','Karnataka','Mizoram','Pondicherry','Meghalaya','AndamanandNicobarIslands','HimachalPradesh'])]
sns.catplot(y='phone_brand',kind='count',height=18,order=topbrands.phone_brand.value_counts().index[:11],data=topbrands,hue = 'state')

#6 Distribution of Gender for each State, Age Segment and Phone Brand(Consider only 10 Most used Phone Brands)

In [ ]:
fig,ax =plt.subplots(1,3,figsize=(20,8))
plt.title('Distribution of genders across states, age-groups & phone brands(top 10)', loc='left', pad=10)

#plot gender across states
sns.countplot(data = dfunique_users,x = 'gender', hue='state', ax=ax[0], palette="muted")

#plot gender across age groups
sns.countplot(data = dfunique_users,x = 'gender', hue='new_group', ax=ax[1], palette="RdBu")

#plot gender across phone brands
sns.countplot(data = dfunique_users[dfunique_users['phone_brand'].isin(top10_brands)],x = 'gender', hue='phone_brand', ax=ax[2], palette="vlag")

#set axis labels
ax[0].set(xlabel='Gender', ylabel='Number of unique users')
ax[1].set(xlabel='Gender', ylabel='Number of unique users')
ax[2].set(xlabel='Gender', ylabel='Number of unique users')

ax[0].set_xticklabels(['Male', 'Female'])
ax[1].set_xticklabels(['Male', 'Female'])
ax[2].set_xticklabels(['Male', 'Female'])

**Observations**


1.   **Distribution of users by genders across 7 states**
  *   Among the 7 states, AndhraPradesh & Karnataka states have most of the   INSAID Telecom users. For rest of the 5 states, there are negligible number of users.  
  *   Amongst Male users, state of Karnataka is leading followed by the state of AndhraPradesh. Karnataka has approx 4,200 & AP has approx 3,200 male users.
  *   The trend for Female users is similar to that of the Male users. Karnataka leads with approx 2,400 & AP with approx 1,800 female users.


2.   **Distribution of users by genders across age groups**
  *   People between 21 to 30 age bracket seemingly 
*   List item


** Hourly Distribution of Phone Calls

In [ ]:
dfunique_hrs = dfevents_combined[['hour','event_id']]
df_hrcounts = dfunique_hrs.groupby(by=['hour'], as_index=False).agg({'event_id': pd.Series.nunique})
df_hrcounts.rename(columns={'event_id': 'callcount'}, inplace=True)

In [ ]:
sns.set(font_scale=1.2)
df_hrcounts.set_index('hour')['callcount'].plot(figsize=(15, 6), linewidth=2.5, color='maroon')
plt.xlabel("Hour", labelpad=12)
plt.ylabel("Call Count", labelpad=12)
plt.title("Hourly Call Counts", y=1.02, fontsize=18);
plt.xticks(ticks=df_hrcounts['hour'])
plt.show()

### Observation:
> - **Majority of the users use phone at 10th hour.**


## Statewise User Count on Map

In [ ]:
dfuuser = dfunique_users.groupby(by=['state'], as_index=False).agg({'device_id': pd.Series.nunique})
dfuuser.rename(columns={'device_id': 'userscount'}, inplace=True)
dfuuser['latitude'] = 0.0 
dfuuser['longitude'] = 0.0
dfuuser['state'][0]= 'Andaman and Nicobar'
dfuuser['state'][1]= 'Andhra Pradesh'
dfuuser['state'][2]= 'Himachal Pradesh'
dfuuser['state'][6]= 'Puducherry'

In [ ]:
dfuuser.head()

In [ ]:
dfuuser.head(10)

In [ ]:
from geopy.geocoders import GoogleV3
googleGeo = GoogleV3('AIzaSyASL2lBX4SQTr6533dDb_62lhW7PDYzgIY')

# create a geocoded list containing geocode objects
geocoded = []

for address in dfuuser['state']:  # assumes dfuuser is a pandas df
    geocoded.append(googleGeo.geocode(address + ",IN"))  # geocode function returns a geocoded object

# append geocoded list to mydata
dfuuser['geocoded'] = geocoded

# if you want to split our your lat and long then do
dfuuser['latitude'] = dfuuser['geocoded'].apply(lambda x: x.latitude)
dfuuser['longitude'] = dfuuser['geocoded'].apply(lambda x: x.longitude)

In [ ]:
import plotly.express as px
import plotly.graph_objs as go
#px.set_mapbox_access_token("pk.eyJ1IjoidmlzaGFsamc1NTgxIiwiYSI6ImNrY2oxcXI3NTE0c2MzNHFxOXdkNjk5YjcifQ.Dpa8vgpwKX75BPLfgO8ouA")
mapbox_access_token= "pk.eyJ1IjoidmlzaGFsamc1NTgxIiwiYSI6ImNrY2oxcXI3NTE0c2MzNHFxOXdkNjk5YjcifQ.Dpa8vgpwKX75BPLfgO8ouA"

#df = px.data.carshare()
fig = px.scatter_mapbox( dfuuser, lat="latitude", lon="longitude",color="userscount", size="userscount",
        color_continuous_scale=px.colors.diverging.RdYlGn, size_max=50, zoom=10,labels=dfuuser['state'],text=dfuuser['state'],height = 700)

fig.update_layout(
    hovermode='closest',
    mapbox=dict(
        accesstoken=mapbox_access_token,
        bearing=1,
        center=go.layout.mapbox.Center(
            lat=21.1458,
            lon=79.0882
        ),
        pitch=0,
        zoom=4,
        # default map style
        #"open-street-map", "carto-positron", "carto-darkmatter", "stamen-terrain", "stamen-toner" or "stamen-watercolor"
        #"basic", "streets", "outdoors", "light", "dark", "satellite", or "satellite-streets"
        style = 'basic',
        
        
    )
)

fig.show()